modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngine.java

```diff
@@ -153,80 +149,48 @@ public CompletableFuture<Void> onUpdate(WatchEvent evt) {
+                                tableConfig,
+                                zoneConfig,
+                                filteredDataNodes,
+                                evt.entryEvent().newEntry().revision(),
+                                metaStorageManager
                         );
 
-                        if (filteredDataNodes.isEmpty()) {
-                            continue;
-                        }
-
-                        if (zoneId == tableZoneId) {
-                            TableConfiguration tableCfg = tables.get(tableView.name());
-
-                            int tableId = tableCfg.id().value();
-
-                            CompletableFuture<List<Set<Assignment>>> tableAssignmentsFut = tableAssignments(
-                                    metaStorageManager,
-                                    tableId,
-                                    distributionZoneConfiguration.partitions().value()
-                            );
-
-                            tableAssignmentsFut.thenAccept(tableAssignments -> {
-
-                                for (int part = 0; part < distributionZoneConfiguration.partitions().value(); part++) {
-
-                                    TablePartitionId replicaGrpId = new TablePartitionId(tableId, part);
-
-                                    int replicas = distributionZoneConfiguration.replicas().value();
-
-                                    int partId = part;
-
-                                    updatePendingAssignmentsKeys(
-                                            tableView.name(),
-                                            replicaGrpId,
-                                            filteredDataNodes,
-                                            replicas,
-                                            evt.entryEvent().newEntry().revision(),
-                                            metaStorageManager,
-                                            partId,
-                                            tableAssignments.isEmpty() ? Collections.emptySet() : tableAssignments.get(partId)
-                                    ).exceptionally(e -> {
-                                        LOG.error(
-                                                "Exception on updating assignments for [table={}, partition={}]", e, tableView.name(),
-                                                partId
-                                        );
-
-                                        return null;
-                                    });
+                        Set<Throwable> exceptions = newSetFromMap(new ConcurrentHashMap<>());
+                        for (int partId = 0; partId < partitionFutures.length; partId++) {
+                            int finalPartId = partId;
 
+                            partitionFutures[partId].exceptionally(e -> {
+                                if (exceptions.add(e)) {
```

Comment:
Do we use exceptions set anywhere?

Reply:
I added a couple of comments to clarify the situation